### PR TITLE
feat: implement a method for oauth2 loginWithPopup

### PR DESCRIPTION
**Pull request by sherzod-bakhodirov**
### 📦 Pull Request

https://magiclink.atlassian.net/browse/PDEEXP-1920

### ✅ Fixed Issues

- [List any fixed issues here like: Fixes #XXXX]

### 🚨 Test instructions

https://github.com/user-attachments/assets/08b9138f-a62b-495a-b303-f242c424c832

[Describe any additional context required to test the PR/feature/bug fix.]

### ⚠️ Don't forget to add a [semver](https://semver.org/) label! 
Please 🚨 **ONLY ADD ONE** 🚨 of the following labels, failing to do so may lead to adverse versioning of your changes when published:
- `patch`: Bug Fix?
- `minor`: New Feature?
- `major`: Breaking Change?
- `skip-release`: It's unnecessary to publish this change.

### Special Note
Please avoid adding any of the `Priority` labels as they conflict with the labels above ☝️

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @magic-ext/algorand@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/aptos@11.13.0-canary.820.11387417874.0
  npm install @magic-ext/avalanche@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/bitcoin@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/conflux@21.13.0-canary.820.11387417874.0
  npm install @magic-ext/cosmos@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/ed25519@19.13.0-canary.820.11387417874.0
  npm install @magic-ext/farcaster@0.13.0-canary.820.11387417874.0
  npm install @magic-ext/flow@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/gdkms@11.13.0-canary.820.11387417874.0
  npm install @magic-ext/harmony@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/icon@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/kadena@0.3.0-canary.820.11387417874.0
  npm install @magic-ext/near@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/oauth@22.13.0-canary.820.11387417874.0
  npm install @magic-ext/oauth2@9.13.0-canary.820.11387417874.0
  npm install @magic-ext/oidc@11.13.0-canary.820.11387417874.0
  npm install @magic-ext/polkadot@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/react-native-bare-oauth@25.14.0-canary.820.11387417874.0
  npm install @magic-ext/react-native-expo-oauth@25.14.0-canary.820.11387417874.0
  npm install @magic-ext/solana@25.14.0-canary.820.11387417874.0
  npm install @magic-ext/sui@0.14.0-canary.820.11387417874.0
  npm install @magic-ext/taquito@20.13.0-canary.820.11387417874.0
  npm install @magic-ext/terra@20.13.0-canary.820.11387417874.0
  npm install @magic-ext/tezos@23.13.0-canary.820.11387417874.0
  npm install @magic-ext/webauthn@22.13.0-canary.820.11387417874.0
  npm install @magic-ext/zilliqa@23.13.0-canary.820.11387417874.0
  npm install @magic-sdk/commons@24.13.0-canary.820.11387417874.0
  npm install @magic-sdk/pnp@22.13.0-canary.820.11387417874.0
  npm install @magic-sdk/provider@28.13.0-canary.820.11387417874.0
  npm install @magic-sdk/react-native-bare@29.14.0-canary.820.11387417874.0
  npm install @magic-sdk/react-native-expo@29.14.0-canary.820.11387417874.0
  npm install @magic-sdk/types@24.11.0-canary.820.11387417874.0
  npm install magic-sdk@28.13.0-canary.820.11387417874.0
  # or 
  yarn add @magic-ext/algorand@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/aptos@11.13.0-canary.820.11387417874.0
  yarn add @magic-ext/avalanche@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/bitcoin@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/conflux@21.13.0-canary.820.11387417874.0
  yarn add @magic-ext/cosmos@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/ed25519@19.13.0-canary.820.11387417874.0
  yarn add @magic-ext/farcaster@0.13.0-canary.820.11387417874.0
  yarn add @magic-ext/flow@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/gdkms@11.13.0-canary.820.11387417874.0
  yarn add @magic-ext/harmony@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/icon@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/kadena@0.3.0-canary.820.11387417874.0
  yarn add @magic-ext/near@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/oauth@22.13.0-canary.820.11387417874.0
  yarn add @magic-ext/oauth2@9.13.0-canary.820.11387417874.0
  yarn add @magic-ext/oidc@11.13.0-canary.820.11387417874.0
  yarn add @magic-ext/polkadot@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/react-native-bare-oauth@25.14.0-canary.820.11387417874.0
  yarn add @magic-ext/react-native-expo-oauth@25.14.0-canary.820.11387417874.0
  yarn add @magic-ext/solana@25.14.0-canary.820.11387417874.0
  yarn add @magic-ext/sui@0.14.0-canary.820.11387417874.0
  yarn add @magic-ext/taquito@20.13.0-canary.820.11387417874.0
  yarn add @magic-ext/terra@20.13.0-canary.820.11387417874.0
  yarn add @magic-ext/tezos@23.13.0-canary.820.11387417874.0
  yarn add @magic-ext/webauthn@22.13.0-canary.820.11387417874.0
  yarn add @magic-ext/zilliqa@23.13.0-canary.820.11387417874.0
  yarn add @magic-sdk/commons@24.13.0-canary.820.11387417874.0
  yarn add @magic-sdk/pnp@22.13.0-canary.820.11387417874.0
  yarn add @magic-sdk/provider@28.13.0-canary.820.11387417874.0
  yarn add @magic-sdk/react-native-bare@29.14.0-canary.820.11387417874.0
  yarn add @magic-sdk/react-native-expo@29.14.0-canary.820.11387417874.0
  yarn add @magic-sdk/types@24.11.0-canary.820.11387417874.0
  yarn add magic-sdk@28.13.0-canary.820.11387417874.0
  ```
</details>

### CODE REVIEW

File: packages/@magic-ext/oauth2/src/index.ts [hunk 1]
```diff
@@ -64,6 +67,65 @@ export class OAuthExtension extends Extension.Internal<'oauth2'> {
...
+
+      const width = 448;
+      const height = 568;
+      const left = window.screenLeft + (window.outerWidth / 2 - width / 2);
+      const top = window.screenTop + window.outerHeight * 0.15;
+      const jwt = await createJwt();
+
+      // try to open the pop up.
+      popup = window.open(
+        `https://auth.magic.link/oauth2/popup/start/${configuration.provider}?dpop=${jwt}&magicApiKey=${this.sdk.apiKey}`,
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/index.ts:89]: Recommend to Use qs to construct query string. It's cleaner that way

File: packages/@magic-ext/oauth2/src/index.ts [hunk 2]
```diff
@@ -64,6 +67,65 @@ export class OAuthExtension extends Extension.Internal<'oauth2'> {
...
+      );
+
+      // see if the popup was blocked and throw an error accordingly
+      if (!popup) {
+        reject(new Error('Failed to open popup window'));
+        return;
+      }
+
+      const messageListener = (event: MessageEvent) => {
+        if (event.data.msgType !== MagicIncomingWindowMessage.MAGIC_POPUP_RESPONSE) return;
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/index.ts:101]: Please add the logic to check event origin to make sure the message is from window we owned.

File: packages/@magic-ext/oauth2/src/index.ts [hunk 3]
```diff
@@ -64,6 +67,65 @@ export class OAuthExtension extends Extension.Internal<'oauth2'> {
...
+          } else {
+            resolve(event.data.payload);
+          }
+        }
+      };
+
+      window.addEventListener('message', messageListener);
+
+      const checkPopupClosed = setInterval(() => {
+        if (popup?.closed) {
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/index.ts:120]: I think there might be better way to check if the popup is closed. Could you check how passport sdk does this?

File: packages/@magic-ext/oauth2/src/types.ts [hunk 4]
```diff
@@ -111,3 +112,27 @@ export enum OAuthErrorCode {
   ServerError = 'server_error',
   TemporarilyUnavailable = 'temporarily_unavailable',
 }
+
+export interface OAuthPopupConfiguration {
+  provider: OAuthProvider;
+}
+
+export interface OAuthPopupResult {
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/types.ts:120]: Is it possible to align this result with "oauth-redirect-result"?

File: packages/@magic-ext/oauth2/src/index.ts (hunk 3 above)
Comment (sherzod-bakhodirov) [reply packages/@magic-ext/oauth2/src/index.ts:120]: I think this is the way we should check, passport also does it 
https://github.com/magiclabs/magic-passport/blob/cd21baf23ae60f7fa34639e34e425acb88e517cc/src/modules/base.ts#L56

**Pull request closed without merging**